### PR TITLE
Add .devcontainer

### PR DESCRIPTION
**Pull request by andrewvc**
## Proposed commit message

Adds `.devcontainer`, a standard way to describe development containers. See the [VSCode](https://code.visualstudio.com/docs/devcontainers/containers) documentation and the [.devcontainer spec](https://containers.dev/) for more info. You can use devcontainers with other IDEs like `intellij`, see [supported tools](https://containers.dev/supporting) for more info.  If you haven't used dev containers before, they're simple to get started. If this file is present in your container VSCode will ask you on opening the project whether you want to use the container. If so, it will build the container and you can go from there.

This file can be ignored by those who wish to develop on their native system, but is highly useful for those (like myself) who occasionally:

1. Need to develop linux specific things that depend on linux kernel features
2. Want to not have to setup tools on a new box
3. Want to share the burden of setting up a dev stack.

This devcontainer is somewhat barebones for beats dev, but does include:

1. golang
2. Mage
3. node (for synthetics)
4. docker (using host docker)
5. ZSH as default shell, with other helpful utils

The goal of this commit is to be minimal, and not include every useful thing for beats dev, follow-up PRs could include more specific functionality.

## Checklist

~- [ ] My code follows the style guidelines of this project~
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
~- [ ] I have made corresponding change to the default configuration files~
~- [ ] I have added tests that prove my fix is effective or that my feature works~
~- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.~

## How to test this PR locally

Open the project in VS Code, use the devcontainer.

## Related issues

- 

## Screenshots

image

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/go
+// If you're wondering how to get SSH working for git pushes, make sure to ssh-add ~/.ssh/<YOURPRIVATEKEY> on your host system
+// to make sure the ssh agent has your key
+{
+	"name": "Go",
+	// Or use a Dockerfile or Docker Compose file. More info: https://containers.dev/guide/dockerfile
+	"image": "mcr.microsoft.com/devcontainers/go:1-1.21-bullseye",
```
Comment (ycombinator) [inline .devcontainer/devcontainer.json:8]: Would it make sense to have the Go version string here be managed by [`updatecli`](https://www.updatecli.io/)? I think what that means specifically is that you'll want to add a new target in this section: https://github.com/elastic/beats/blob/3161fc0cd2e4250508ab4a40bd2debdf0883bdb7/.ci/bump-golang.yml#L86

Comment (andrewvc) [reply .devcontainer/devcontainer.json:8]: Yes! Would you be able to add that as a commit?

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/go
+// If you're wondering how to get SSH working for git pushes, make sure to ssh-add ~/.ssh/<YOURPRIVATEKEY> on your host system
+// to make sure the ssh agent has your key
+{
+	"name": "Go",
```
Comment (ycombinator) [inline .devcontainer/devcontainer.json:6]: Nit: could we rename this to something slightly more specific, e.g.:
```suggestion
	"name": "Beats Development Container",
```

File: .devcontainer/devcontainer.json (hunk 1 above)
Comment (ycombinator) [reply .devcontainer/devcontainer.json:8]: What is the difference between `go:1-1.21-bullseye` and `go:1.21-bullseye`? I found https://github.com/devcontainers/images/tree/main/src/go#using-this-image but I couldn't understand the effect of adding or leaving out the extra `1-` at the start of the tag.

Comment (ycombinator) [reply .devcontainer/devcontainer.json]: > Yes! Would you be able to add that as a commit?

Done via 9a897f1f3796e06bc533c2e9151fbd62531438d2 and 29fdddf7a333eca9bdcb953e80212e3f95268c72.

**Comment by andrewvc** [comment]
Thanks for the improvements @ycombinator !

**Review by ycombinator — COMMENTED** [review]
This change LGTM but I do have one question: https://github.com/elastic/beats/pull/36422#discussion_r1306234188

**Comment by ycombinator** [comment]
Removing the `backport-7.17` label as that branch does not use `updatecli` to automatically maintain Go versions; instead it uses a bespoke script: https://github.com/elastic/beats/blob/7.17/.ci/bump-go-release-version.sh.

**Review by ycombinator — APPROVED** [review]
LGTM.

**Pull request merged**